crates/bevy_ecs/src/query/iter.rs

```diff
@@ -157,6 +159,640 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIter<'w, 's, D, F> {
+    /// # struct Height(usize);
+    /// #
+    /// # #[derive(Component, PartialEq, Eq, PartialOrd, Ord)]
+    /// # struct ParentEntity(Entity);
+    /// #
+    /// # #[derive(Component, Clone, Copy)]
+    /// # struct ChildPartCount(usize);
+    /// #
+    /// # impl Deref for ChildPartCount {
+    /// #     type Target = usize;
+    /// #
+    /// #     fn deref(&self) -> &Self::Target {
+    /// #         &self.0
+    /// #     }
+    /// # }
+    /// # let mut world = World::new();
+    /// // We can ensure that a query always returns in the same order.
+    /// fn system_1(query: Query<(Entity, &PartIndex)>) {
+    ///     let parts: Vec<(Entity, &PartIndex)> = query.iter().sort::<&PartIndex>().collect();
+    /// }
+    ///
+    /// // We can freely rearrange query components in the key.
+    /// fn system_2(query: Query<(&Length, &Width, &Height), With<PartMarker>>) {
+    ///     for (length, width, height) in query.iter().sort::<(&Height, &Length, &Width)>() {
+    ///         println!("height: {height:?}, width: {width:?}, length: {length:?}")
+    ///     }
+    /// }
+    ///
+    /// // We can sort by Entity without including it in the original Query.
+    /// // Here, we match iteration orders between query iterators.
+    /// fn system_3(
+    ///     part_query: Query<(&PartValue, &ParentEntity)>,
+    ///     mut parent_query: Query<(&ChildPartCount, &mut ParentValue)>,
+    /// ) {
+    ///     let part_values = &mut part_query
+    ///         .into_iter()
+    ///         .sort::<&ParentEntity>()
+    ///         .map(|(&value, parent_entity)| *value);
+    ///
+    ///     for (&child_count, mut parent_value) in parent_query.iter_mut().sort::<Entity>() {
+    ///         **parent_value = part_values.take(*child_count).sum();
+    ///     }
+    /// }
+    /// #
+    /// # let mut schedule = Schedule::default();
+    /// # schedule.add_systems((system_1, system_2, system_3));
+    /// # schedule.run(&mut world);
+    /// ```
+    pub fn sort<L: QueryData + 'w>(
+        self,
+    ) -> QuerySortedIter<
+        'w,
+        's,
+        D,
+        F,
+        impl ExactSizeIterator<Item = Entity> + DoubleEndedIterator + FusedIterator + 'w,
+    >
+    where
+        L::Item<'w>: Ord,
+    {
+        // On the first successful iteration of `QueryIterationCursor`, `archetype_entities` or `table_entities`
+        // will be set to a non-zero value. The correctness of this method relies on this.
+        // I.e. this sort method will execute if and only if `next` on `QueryIterationCursor` of a
+        // non-empty `QueryIter` has not yet been called. When empty, this sort method will not panic.
+        if !self.cursor.archetype_entities.is_empty() || !self.cursor.table_entities.is_empty() {
+            panic!("it is not valid to call sort() after next()")
+        }
+
+        let world = self.world;
+
+        let query_lens_state = self
+            .query_state
```

Comment:
utilizing a readonly query state would be more suitable here

Reply:
you're right, I'll adjust the `L` trait bound to `ReadOnlyQueryData`, what would've been optimal was leaving it as `QueryData` when the underlying iterator is mutable, but that would require specialization.